crates/locutus-router/src/lib.rs

```diff
@@ -1,178 +1,399 @@
+            failure_estimator: IsotonicEstimator::new(failure_outcomes, EstimatorType::Positive),
+            // Negative because we expect transfer rate to decrease as distance increases
+            transfer_rate_estimator: IsotonicEstimator::new(
+                transfer_rates,
+                EstimatorType::Negative,
+            ),
+            mean_transfer_size,
         }
     }
 
-    /// Adds a new event to the estimator.
-    pub fn add_event(&mut self, event: RoutingEvent) {
-        let point = Point::new(
-            event
-                .peer_location
-                .distance(&event.contract_location)
-                .into(),
-            event.result,
-        );
-
-        self.global_regression.add_points(&[point]);
-
-        self.peer_regressions
-            .entry(event.peer)
-            .or_insert_with(|| IsotonicRegression::new_ascending(&[point]))
-            .add_points(&[point]);
+    pub fn add_event(&mut self, event: RouteEvent) {
+        match event.outcome {
+            RouteOutcome::Success {
+                time_to_response_start,
+                payload_size,
+                payload_transfer_time,
+            } => {
+                self.response_start_time_estimator.add_event(IsotonicEvent {
+                    peer: event.peer,
+                    contract_location: event.contract_location,
+                    result: time_to_response_start.as_secs_f64(),
+                });
+                self.failure_estimator.add_event(IsotonicEvent {
+                    peer: event.peer,
+                    contract_location: event.contract_location,
+                    result: 0.0,
+                });
+                let transfer_rate_event = IsotonicEvent {
+                    peer: event.peer,
+                    contract_location: event.contract_location,
+                    result: payload_size as f64 / payload_transfer_time.as_secs_f64(),
+                };
+                self.mean_transfer_size.add(payload_size as f64);
+
+                self.transfer_rate_estimator.add_event(transfer_rate_event);
+            }
+            RouteOutcome::Failure => {
+                self.failure_estimator.add_event(IsotonicEvent {
+                    peer: event.peer,
+                    contract_location: event.contract_location,
+                    result: 1.0,
+                });
+            }
+        }
     }
 
-    pub fn estimate_retrieval_time(&self, peer: PeerId, distance: f64) -> Option<f64> {
-        if let Some(regression) = self.peer_regressions.get(&peer) {
-            Some(regression.interpolate(distance))
-        } else if self.global_regression.len() > MIN_PEER_POINTS_FOR_REGRESSION {
-            Some(self.global_regression.interpolate(distance))
+    pub fn select_peer<I>(&self, peers: I, contract_location: Location) -> Option<PeerKeyLocation>
+    where
+        I: IntoIterator<Item = PeerKeyLocation>,
```

Comment:
Is possible we can handle a reference here instead 
`select_peer<'a>(...) -> Option<>&'a PeerKeyLocation> where I: IntoIterator<Item= &'a PeerKeyLocation> `

Reply:
Yes, have made the change and a few other similar changes to pass references rather than owned structs.